circuits/cpp/src/aztec3/circuits/kernel/public/common.hpp

```diff
@@ -0,0 +1,178 @@
+}
+
+template <typename KernelInput>
+void validate_state_reads(DummyComposer& composer, KernelInput const& public_kernel_inputs)
+{
+    // Validates that state reads correspond to the provided membership witnesses
+    const auto& reads = public_kernel_inputs.public_call.public_call_data.call_stack_item.public_inputs.state_reads;
+    const auto& contract_address = public_kernel_inputs.public_call.public_call_data.call_stack_item.contract_address;
+    for (size_t i = 0; i < STATE_READS_LENGTH; ++i) {
+        const auto& state_read = reads[i];
+        if (state_read.is_empty()) {
+            continue;
+        }
+        const auto& sibling_path = public_kernel_inputs.public_call.state_reads_sibling_paths[i].sibling_path;
+        const NT::fr leaf_value = compute_public_data_tree_value<NT>(state_read.current_value);
+        const NT::fr leaf_index = compute_public_data_tree_index<NT>(contract_address, state_read.storage_slot);
+        const std::string msg = format("validate_state_reads, index ", i, " leaf value ", leaf_value);
+        check_membership(composer,
+                         leaf_value,
+                         leaf_index,
+                         sibling_path,
+                         public_kernel_inputs.public_call.public_data_tree_root,
+                         msg);
+    }
+};
+
+template <typename KernelInput>
+void validate_state_transitions(DummyComposer& composer, KernelInput const& public_kernel_inputs)
+{
+    // Validates that the old value of state transitions correspond to the provided membership witnesses
+    const auto& transitions =
+        public_kernel_inputs.public_call.public_call_data.call_stack_item.public_inputs.state_transitions;
+    const auto& contract_address = public_kernel_inputs.public_call.public_call_data.call_stack_item.contract_address;
+    for (size_t i = 0; i < STATE_TRANSITIONS_LENGTH; ++i) {
+        const auto& state_transition = transitions[i];
+        if (state_transition.is_empty()) {
+            continue;
+        }
+        const auto& sibling_path = public_kernel_inputs.public_call.state_transitions_sibling_paths[i].sibling_path;
+        const NT::fr leaf_value = compute_public_data_tree_value<NT>(state_transition.old_value);
+        const NT::fr leaf_index = compute_public_data_tree_index<NT>(contract_address, state_transition.storage_slot);
+        const std::string msg = format("validate_state_transitions, index ", i, " leaf value ", leaf_value);
+        check_membership(composer,
+                         leaf_value,
+                         leaf_index,
+                         sibling_path,
+                         public_kernel_inputs.public_call.public_data_tree_root,
+                         msg);
+    }
+};
+
+template <typename KernelInput>
+void validate_this_public_call_stack(DummyComposer& composer, KernelInput const& public_kernel_inputs)
+{
+    // Ensures that the stack of pre-images corresponds to the call stack
```

Comment:
For all of these, try throwing these comments in a doxygen-style comment with at least a `@brief`. So here, move this before the function signature and format it as:

```
/** 
 * @brief Ensures that the stack of pre-images corresponds to the call stack
 */
template <typename KernelInput>
void validate_this_public_call_stack(DummyComposer& composer, KernelInput const& public_kernel_inputs)
{
```

We intend to auto-generate docs in the future.